onadata/libs/serializers/metadata_serializer.py

```diff
@@ -101,23 +96,91 @@ def create(self, validated_data):
             from_kpi=from_kpi,
         )
 
-    def _validate_data_file_type(self, data_file_type, data_file, data_value):
+    def _validate_data_value(self, attrs):
+        if self.instance and self.instance.pk:
+            # ToDo , Should we allow PATCHing the object with different file?
+            attrs['data_file'] = attrs.get(
+                'data_file', self.instance.data_file
+            )
+            attrs['data_type'] = attrs.get(
+                'data_type', self.instance.data_type
+            )
+            try:
+                attrs['data_value']
+            except KeyError:
+                attrs['data_value'] = self.instance.data_value
+                return attrs
 
-        data_value = (
-            data_file.name if data_file else data_value
-        )
-        allowed_types = settings.SUPPORTED_MEDIA_UPLOAD_TYPES
+        data_value = attrs.get('data_value')
+
+        if attrs.get('data_type') == 'media' and attrs.get('data_file') is None:
+            URLValidator(message=_("Invalid url %s." % data_value))(data_value)
 
-        if not data_file_type:
-            data_file_type = (
-                data_file.content_type
-                if data_file and data_file.content_type in allowed_types
-                else mimetypes.guess_type(data_value)[0]
+        if not data_value:
+            raise serializers.ValidationError(
+                {'data_value': _('This field is required.')}
             )
 
-        if data_file_type not in allowed_types:
+        return attrs
+
+    def _validate_data_file_type(self, attrs):
+
+        allowed_types = settings.SUPPORTED_MEDIA_UPLOAD_TYPES
+
+        data_file = attrs.get('data_file')
+        data_value = attrs.get('data_value')
```

Comment:
We've now set these values in `attrs` so we can do:
```python
data_file = attrs['data_file']
data_value = attrs['data_value']
```

Reply:
I agree with you for `data_value` but not for `data_file`. 
On `POST` you can send only the `data_value` (e.g. remote URL) , so `data_file` would be `None` and would raise an `KeyError`.